sympy/ntheory/generate.py

```diff
@@ -39,16 +50,32 @@ class Sieve:
     """
 
     # data shared (and updated) by all Sieve instances
-    _list = _array('l', [2, 3, 5, 7, 11, 13])
+    def __init__(self):
+        self._n = 6
+        self._list = _aset(2, 3, 5, 7, 11, 13) # primes
+        self._tlist = _aset(0, 1, 1, 2, 2, 4) # totient
+        self._mlist = _aset(0, 1, -1, -1, 0, -1) # mobius
 
     def __repr__(self):
-        return "<Sieve with %i primes sieved: 2, 3, 5, ... %i, %i>" % \
-            (len(self._list), self._list[-2], self._list[-1])
+        return ("<Sieve with \n%s (%i): %i, %i, %i, ... %i, %i\n"
+             "%s (%i): %i, %i, %i, ... %i, %i\n"
+             "%s (%i): %i, %i, %i, ... %i, %i>") % (
+             'primes', len(self._list),
+                 self._list[0], self._list[1], self._list[2],
+                 self._list[-2], self._list[-1],
+             'totient', len(self._tlist),
+                 self._tlist[0], self._tlist[1],
+                 self._tlist[2], self._tlist[-2], self._tlist[-1],
+             'mobius', len(self._mlist),
+                 self._mlist[0], self._mlist[1],
+                 self._mlist[2], self._mlist[-2], self._mlist[-1])
 
     def _reset(self):
         """Return sieve to its initial state for testing purposes.
         """
-        self._list = self._list[:6]
+        self._list = self._list[:self._n]
+        self._tlist = self._tlist[:self._n]
+        self._mlist = self._mlist[:self._n]
```

Comment:
args for `prime=True, totient=True, mobius=True` might be useful for `_reset`.

Reply:
yes. Also, assert in `__init__` that all lists have length `_n`

Reply:
Yes, I thought so. Perhaps:

```python
def _reset(self, **which):
    """reset all caches to initial state (default). To reset only one or more, indicate
    which to reset with `prime=True`, `totient=True` and/or `mobius=True`."""
    if not which:
        p, m, t = [True]*3
    else:
        ok = 'prime totient mobius'.split()
        for k in which:
            if k not in ok:
                raise ValueError('keyword must be one of %s' % ok)
        p, t, m = [which.get(i, False) for i in ok]
        if p:
            self._list = self._list[:self._n]
        if t:
            self._tlist = self._tlist[:self._n]
        if m:
            self._mlist = self._mlist[:self._n]
```